In [25]:
import numpy as np
import pandas as pd

In [63]:
import seaborn as sns
import matplotlib.pyplot as plt

%matplotlib inline

In [64]:
from sklearn.decomposition import PCA

In [67]:
df=pd.read_csv('loan.csv', sep=',', header=0)

data = df.drop(
    df.columns[0], 
    axis=1)

# Drop rows with missing column data
data = data.dropna()


## Exploratory Data Analysis

In [68]:
data.head()

SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines  age  \
0                 1                              0.766127   45   
1                 0                              0.957151   40   
2                 0                              0.658180   38   
3                 0                              0.233810   30   
4                 0                              0.907239   49   

   NumberOfTime30-59DaysPastDueNotWorse  DebtRatio  MonthlyIncome  \
0                                     2   0.802982         9120.0   
1                                     0   0.121876         2600.0   
2                                     1   0.085113         3042.0   
3                                     0   0.036050         3300.0   
4                                     1   0.024926        63588.0   

   NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
0                               13                        0   
1                                4                        0   
2                                2                        1   
3                                5                        0   
4                                7                        0   

   NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
0                             6                                     0   
1                             0                                     0   
2                             0                                     0   
3                             0                                     0   
4                             1                                     0   

   NumberOfDependents  
0                 2.0  
1                 1.0  
2                 0.0  
3                 0.0  
4                 0.0

## Drop nan and header

In [69]:
data.describe().transpose()


count         mean           std  \
SeriousDlqin2yrs                      120269.0     0.069486      0.254280   
RevolvingUtilizationOfUnsecuredLines  120269.0     5.899873    257.040685   
age                                   120269.0    51.289792     14.426684   
NumberOfTime30-59DaysPastDueNotWorse  120269.0     0.381769      3.499234   
DebtRatio                             120269.0    26.598777    424.446457   
MonthlyIncome                         120269.0  6670.221237  14384.674215   
NumberOfOpenCreditLinesAndLoans       120269.0     8.758475      5.172835   
NumberOfTimes90DaysLate               120269.0     0.211925      3.465276   
NumberRealEstateLoansOrLines          120269.0     1.054519      1.149273   
NumberOfTime60-89DaysPastDueNotWorse  120269.0     0.187829      3.447901   
NumberOfDependents                    120269.0     0.851832      1.148391   

                                      min          25%          50%  \
SeriousDlqin2yrs                      0.0     0.000000     0.000000   
RevolvingUtilizationOfUnsecuredLines  0.0     0.035084     0.177282   
age                                   0.0    40.000000    51.000000   
NumberOfTime30-59DaysPastDueNotWorse  0.0     0.000000     0.000000   
DebtRatio                             0.0     0.143388     0.296023   
MonthlyIncome                         0.0  3400.000000  5400.000000   
NumberOfOpenCreditLinesAndLoans       0.0     5.000000     8.000000   
NumberOfTimes90DaysLate               0.0     0.000000     0.000000   
NumberRealEstateLoansOrLines          0.0     0.000000     1.000000   
NumberOfTime60-89DaysPastDueNotWorse  0.0     0.000000     0.000000   
NumberOfDependents                    0.0     0.000000     0.000000   

                                              75%        max  
SeriousDlqin2yrs                         0.000000        1.0  
RevolvingUtilizationOfUnsecuredLines     0.579428    50708.0  
age                                     61.000000      103.0  
NumberOfTime30-59DaysPastDueNotWorse     0.000000       98.0  
DebtRatio                                0.482559    61106.5  
MonthlyIncome                         8249.000000  3008750.0  
NumberOfOpenCreditLinesAndLoans         11.000000       58.0  
NumberOfTimes90DaysLate                  0.000000       98.0  
NumberRealEstateLoansOrLines             2.000000       54.0  
NumberOfTime60-89DaysPastDueNotWorse     0.000000       98.0  
NumberOfDependents                       2.000000       20.0

### Seperate Target and Outcome Features


In [70]:
from sklearn.feature_extraction import DictVectorizer
from pandas import DataFrame
vec = DictVectorizer()

In [71]:
data = data.to_dict(orient='records')

In [72]:
df_data = vec.fit_transform(data).toarray()
feature_names = vec.get_feature_names()

df_data = DataFrame(
    df_data,
    columns=feature_names)

In [73]:
outcome_feature = df_data['SeriousDlqin2yrs']
target_features = df_data.drop('SeriousDlqin2yrs', axis=1)

### Training and Testing Set


In [74]:
from sklearn.model_selection import train_test_split

In [75]:
X_train, X_test, y_train, y_test = train_test_split(
    target_features, outcome_feature, test_size=0.5, random_state=0)

## Define our model classifier

In [76]:
from sklearn.naive_bayes import GaussianNB
clf = GaussianNB()

In [ ]:
#Train The classier

In [77]:
clf.fit(X_train,y_train)

GaussianNB(priors=None)

In [ ]:
## Validate Classifier


In [78]:
output = clf.predict(X_test)

In [79]:
from sklearn.metrics import confusion_matrix
matrix = confusion_matrix(output, y_test)
score = clf.score(X_test, y_test)
print ("\n accuracy:====> {0} \n".format(score.mean()))
print(matrix)


 accuracy:====> 0.9294088301322025 

[[55798  4094]
 [  151    92]]


The output shows that we have over 92% accuracy with the following error types

55798 true positives
151 true negatives
92 false positives
4094 false negatives


## Save Classifier
With our classifier done we can save it so that we can use it a separate program

In [80]:
from sklearn.externals import joblib
joblib.dump(clf, 'model/loan_default.pkl')

['model/loan_default.pkl']